src/AutoMapper/Execution/TypeMapPlanBuilder.cs

```diff
@@ -64,6 +66,64 @@ public LambdaExpression CreateMapperLambda()
             return Lambda(Block(new[] { _destination }, lambaBody), _source, _initialDestination, _context);
         }
 
+        private void CheckForCycles(HashSet<TypeMap> visitedTypeMaps)
+        {
+            if(_typeMap.PreserveReferences)
+            {
+                return;
+            }
+            bool isRoot;
+            if(visitedTypeMaps == null)
+            {
+                isRoot = true;
+                visitedTypeMaps = new HashSet<TypeMap>();
+            }
+            else
+            {
+                isRoot = false;
+            }
+
+            CheckPropertyMapsForCycles();
+
+            if(isRoot && visitedTypeMaps.Contains(_typeMap))
+            {
+                _typeMap.PreserveReferences = true;
+            }
+
+            void CheckPropertyMapsForCycles()
+            {
+                var propertyTypeMaps =
+                    from propertyTypeMap in
+                    (from pm in _typeMap.GetPropertyMaps() where pm.CanResolveValue() select ResolvePropertyTypeMap(pm))
+                    where propertyTypeMap != null && !propertyTypeMap.PreserveReferences && !visitedTypeMaps.Contains(propertyTypeMap)
+                    select propertyTypeMap;
+                foreach(var propertyTypeMap in propertyTypeMaps)
+                {
+                    visitedTypeMaps.Add(propertyTypeMap);
+                    propertyTypeMap.Seal(_configurationProvider, visitedTypeMaps);
+                }
+            }
+        }
+
+        private TypeMap ResolvePropertyTypeMap(PropertyMap propertyMap)
+        {
+            if(propertyMap.SourceType == null)
+            {
+                return null;
+            }
+            var types = new TypePair(propertyMap.SourceType, propertyMap.DestinationPropertyType);
+            var typeMap = _configurationProvider.ResolveTypeMap(types);
+            if(typeMap == null)
+            {
+                var mapper = _configurationProvider.FindMapper(types) as IObjectMapperInfo;
+                if(mapper != null)

```

Comment:
Yes, it was smth more compact in an earlier version, but for debugging, it was easier to write it more explicit.